### PR TITLE
[DDW-701]: Improve Daedalus log rotation

### PR DESCRIPTION
**Pull request by yakovkaravelov**
This PR implements improved Daedalus log rotation.
[YouTrack Ticket](https://iohk.myjetbrains.com/youtrack/issue/DDW-701)

## Todos
- [x] Preserve 3 old log files and 1 active log file each of them up to specified size limit.
- [x] After active file reaches to size limit and there are already 3 old log files, the oldest old file should be deleted.
- [x] All old log file names should be postfixed by datetime stamp.

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1562275190238900)
- [x] Watch log file directory `~/Library/Application Support/Daedalus/Logs/pub` and ensure there are always up to 3 old log files named `Daedalus.json-{timestamp}` and 1 active log file named `Daedalus.json`. 
---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Comment by IuliaDolishniak** [comment]
Hava tested on Win and Linux and seems working fine 
screen

**Comment by nikolaglumac** [comment]
I am not happy with the file naming and the implementation. Will submit change request on Monday.

**Comment by nikolaglumac** [comment]
This PR is NOT to be reviewed nor tested until this one has been merged: https://github.com/input-output-hk/electron-log-daedalus/pull/2 and new version of `electron-log-daedalus` package released and used in this PR...

File: source/main/utils/setupLogging.js [hunk 1]
```diff
@@ -25,7 +25,9 @@ export const setupLogging = () => {
   log.transports.console.level = isTest ? 'error' : 'info';
   log.transports.rendererConsole.level = isDev ? 'info' : 'error';
   log.transports.file.level = 'debug';
-  log.transports.file.maxSize = 20 * 1024 * 1024;
+  log.transports.file.maxSize = 5 * 1024 * 1024;
```
Comment (tomothespian) [inline source/main/utils/setupLogging.js]: On this way you will set all logs to 5MB even if is original Daedalu.log (not old) and that means that 1xDaedalus.log + 3xDeadelus.log-time() will have sum of 20MB.

Please ensure that Daedalus.log is 20MB and and only old logs are 5MB

Comment (nikolaglumac) [reply source/main/utils/setupLogging.js]: No this is wrong @tomothespian. ALL logs need to be max 5 MB in size.

File: source/main/utils/setupLogging.js [hunk 2]
```diff
@@ -25,7 +25,9 @@ export const setupLogging = () => {
...
+  log.transports.file.maxItems = 4;
```
Comment (daniloprates) [inline source/main/utils/setupLogging.js]: I don't understand why, but setting `maxItems = 4` is actually leaving only 2 old logs. Setting it 5 will leave the correct 3 logs. For a quick test, I set `log.transports.file.maxSize = 100;`

File: source/main/ipc/get-logs.js [hunk 3]
```diff
@@ -25,7 +25,20 @@ export const getLogsChannel: MainIpcChannel<
   GetLogsMainResponse
 > = new MainIpcChannel(GET_LOGS_CHANNEL);
 
-const isFileAllowed = (fileName: string) => includes(ALLOWED_LOGS, fileName);
+const isFileAllowed = (fileName: string) => {
+  let allowed = includes(ALLOWED_LOGS, fileName);
```
Comment (daniloprates) [inline source/main/ipc/get-logs.js]: This part is hard to understand if you don't know the context. Please, change to the following:

```js
const isFileAllowed = (fileName: string) =>
  includes(ALLOWED_LOGS, fileName) || isOldDaedalusLog(fileName);
```

Then:

```js
const isOldDaedalusLog = (fileName: string) => {
  // add the checking here, with proper comments for easy understanding
}
```

File: source/main/utils/setupLogging.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (nikolaglumac) [reply source/main/utils/setupLogging.js]: @daniloprates this is 5MB as size is in bytes.
Still, I agree that this should be added. @yakovkaravelov please see to implement this.

File: source/main/utils/setupLogging.js (hunk 2 above)
Comment (nikolaglumac) [reply source/main/utils/setupLogging.js]: @daniloprates I think that @tomothespian saw similar result when he set the maxSize to a similarly small limit as you did. Please try to set the limit to 1MB. Seems that rotation happens to quickly when we set the limit very low.
Then again, perhaps this points into a bug in the logger code.
@yakovkaravelov can you please investigate this?
We need an explanation of this behavior at least 🙏

File: source/main/ipc/get-logs.js (hunk 3 above)
Comment (nikolaglumac) [reply source/main/ipc/get-logs.js]: Great tip @daniloprates!

File: source/main/utils/setupLogging.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (daniloprates) [reply source/main/utils/setupLogging.js]: oh, that's right! 🤦🏻‍♂️ so @yakovkaravelov:

`log.transports.file.maxSize = 5 * 1024 * 1024; // 5 MB | unit: bytes`

File: source/main/utils/setupLogging.js (hunk 2 above)
Comment (yakovkaravelov) [reply source/main/utils/setupLogging.js]: @daniloprates 
Here, `maxItems = 4` means there are 4 logs including 3 old logs and 1 active log.
Above you are reporting it is leaving only 2 old logs, did you wait until active log exceeds the limit?
if you wait until then, there will be 1 more old file generated, and there will be 3 old logs exactly.
Please test again.

**Comment by yakovkaravelov** [comment]
@daniloprates 
I fixed issues reported by you.
Great feedbacks!
Please review again.
Thanks.

**Comment by yakovkaravelov** [comment]
When to merge? :(

**Comment by nikolaglumac** [comment]
@yakovkaravelov as soon as we get an approval from @gnpf

**Comment by gnpf** [comment]
@nikolaglumac @yakovkaravelov This isn't happening in Linux

**Pull request merged**